### PR TITLE
add ‘No database’ option for gateway when OIDC is selected

### LINKED ISSUE #6763: Can we add option "No database" to generate a gateway without database when OIDC option is selected ?
##### **Overview of the issue**

As OIDC is supported in JHipster generator, can we add option "No database" to generate a gateway without database when OIDC option is selected?

The option "No database" is  already existe for micro service application. It would be nice to add it for the gateway like for the microservices to not generate the corresponding code and configuration.

To run gateway without database we have to manually make some changes and run 'jhipster' again to regenerate it. But it's convenient to make these changes after every upgrade. Below the changes to make:

1. Modify database options in `.yo-rc.json` as follows:
    "databaseType": "no",
    "devDatabaseType": "no",
    "prodDatabaseType": "no",

2. run jhipster

3. Delete some files:
- config/DatabaseConfiguration.java
- config/audit/AuditEventConverter.java
- domain/AbstractAuditingEntity.java
- domain/PersistentAuditEvent.java
- config/liquibase folder and related files

I also noticed that the example to run database docker image is added to READE.me as follow:

For example, to start a no database in a docker container, run:
    docker-compose -f src/main/docker/**no.yml** up -d
To stop it and remove the container, run:
    docker-compose -f src/main/docker/**no.yml** down

##### **Motivation for or Use Case**

Testing microservices with OIDC

##### **Related issues**

#3919

##### **Suggest a Fix**

Doing same  as microservice application + don't generate database example in README.md

##### **JHipster Version(s)**

4.11.1

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "gateway",
    "gitHubOrganization": "",
    "baseName": "GatewayTest",
    "packageName": "io.github.jhipster.application",
    "serverPort": 8080,
    "serviceDiscoveryType": "eureka",
    "authenticationType": "oauth2",
    "uaaBaseName": "../uaa",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "no",
    "devDatabaseType": "no",
    "prodDatabaseType": "no",
    "searchEngine": false,
    "enableSwaggerCodegen": false,
    "messageBroker": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "useSass": false,
    "clientPackageManager": "yarn",
    "testFrameworks": [],
    "enableTranslation": false,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ],
    "clientFramework": "angularX",
    "jhipsterVersion": "4.11.1",
    "packageFolder": "io/github/jhipster/application",
    "jhiPrefix": "jhi"
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

None !

##### **Environment and Tools**

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

git version 2.13.1
node: v8.7.0
npm: 5.5.1
bower: 1.8.2
gulp: [01:39:12] CLI version 1.4.0
yeoman: 2.0.0
yarn: 1.3.2
Docker version 17.09.0-ce, build afdb6d4
docker-compose version 1.16.1, build 6d1ac21

##### **Browsers and Operating System**

MacOS and Google Chrome

- [ x ] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by maznag**
Fix #6763

- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/server/prompts.js [hunk 1]
```diff
@@ -157,7 +157,7 @@ function askForServerSideOpts(meta) {
             }
         },
         {
-            when: response => applicationType === 'microservice' || (response.authenticationType === 'uaa' && applicationType === 'gateway'),
+            when: response => applicationType === 'microservice' || ((response.authenticationType === 'uaa' || response.authenticationType === 'oauth2') && applicationType === 'gateway'),
```
Comment (deepu105) [inline generators/server/prompts.js]: Cassandra is not supported for oauth so may be you should just add the `No database` option to the next block specifc for oauth2

File: generators/server/templates/_README.md [hunk 2]
```diff
@@ -296,13 +296,13 @@ For more information, refer to the [Running tests page][].
 ## Using Docker to simplify development (optional)
 
 You can use Docker to improve your JHipster development experience. A number of docker-compose configuration are available in the [src/main/docker](src/main/docker) folder to launch required third party services.
-For example, to start a <%= prodDatabaseType%> database in a docker container, run:
+For example, to start the JHipster Registry in a docker container, run:
```
Comment (deepu105) [inline generators/server/templates/_README.md]: This is unrelated change why?

File: package-lock.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (deepu105) [inline package-lock.json]: unrelated file change. please revert

File: generators/server/prompts.js (hunk 1 above)
Comment (maznag) [reply generators/server/prompts.js]: OK, I'll do it. thanks

File: generators/server/templates/_README.md (hunk 2 above)
Comment (maznag) [reply generators/server/templates/_README.md]: Quite simply because when the 'no database' option is selected, the following doc is generated that makes no sense: 

For example, to start a **no** database in a docker container, run:
docker-compose -f src/main/docker/**no**.yml up -d

File: package-lock.json (hunk 3 above)
Comment (maznag) [reply package-lock.json]: Sorry, I didn't pay attention. this is my first pull request !

File: generators/server/prompts.js (hunk 1 above)
Comment (mraible) [reply generators/server/prompts.js]: Why is Cassandra not supported with OAuth? I believe the previous implementation required database tables, where as the new implementation does not.

Comment (maznag) [reply generators/server/prompts.js]: No idea for Cassandra!

I made the changes. Can I please submit the changes made by doing this on my branch:
git commit -a
git rebase master -i
git push -f

Comment (maznag) [reply generators/server/prompts.js]: @deepu105 It's done.
@mraible No idea for Cassandra !

Comment (deepu105) [reply generators/server/prompts.js]: @mraible I dont know I just inferred from current code and thought it was from you. But if you say it works then it works :)

File: generators/server/templates/_README.md (hunk 2 above)
Comment (deepu105) [reply generators/server/templates/_README.md]: Same way Registry is not always present so if you choose no registry it won't make sense as well and no registry is more common then no database

Comment (maznag) [reply generators/server/templates/_README.md]: @deepu105  May be better than displaying 'No database' and 'docker/no.yml'

So we can add then a condition to test if database is selected. In this case we display the example for database otherwise we display the example forJHipster Registry or Consul as generally 'No database' option is selected in Microservices architecture (gateway or microservice app).

What do you think about that? if you agree I change and I commit.

Comment (deepu105) [reply generators/server/templates/_README.md]: Yes condition is better, if no DB selected then no need to show the section at all

**Comment by pascalgrimaud** [comment]
@maznag : congrats for your 1st contribution!

**Comment by maznag** [comment]
@pascalgrimaud thanks :-)

**Comment by jdubois** [comment]
Wait, what? This is already done???? You rock @maznag !!! Thanks!

**Comment by pascalgrimaud** [comment]
Some comments:
- audits page need to be excluded, like Cassandra:
    - https://github.com/jhipster/generator-jhipster/blob/master/generators/client/files-angular.js#L324-L334
    - https://github.com/jhipster/generator-jhipster/blob/master/generators/client/files-angular.js#L485-L490
- same in the navbar:  https://github.com/jhipster/generator-jhipster/blob/master/generators/client/templates/angular/src/main/webapp/app/layouts/navbar/_navbar.component.html#L103-L110

I didn't check everywhere else.

**Comment by maznag** [comment]
Yes. Also, some i18n related files like audits.json.

**Comment by maznag** [comment]
@pascalgrimaud You can check the last commit; all related audit infos are excluded when ‘No database’ option is selected.

**Comment by pascalgrimaud** [comment]
@maznag : Travis build failed, for `yarn test`. Here the logs from Travis:

```
EJS Lint
    1) EJS templates are valid
  0 passing (630ms)
  1 failing
  1) EJS Lint
       EJS templates are valid:
      AssertionError [ERR_ASSERTION]: Unexpected token (9:5) in generators/client/templates/react/src/main/webapp/robots.txt
      + expected - actual
      -false
      +true
```

File: generators/client/templates/react/src/main/webapp/robots.txt [hunk 4]
```diff
@@ -4,9 +4,7 @@ User-agent: *
 Disallow: /api/account
 Disallow: /api/account/change-password
 Disallow: /api/account/sessions
-<%_ if (databaseType !== 'no' && databaseType !== 'cassandra') ( _%>
```
Comment (pascalgrimaud) [inline generators/client/templates/react/src/main/webapp/robots.txt]: The error was `(` instead of `{`
I didn't see it in my previous comment because I was on my phone

**Comment by pascalgrimaud** [comment]
LGTM, it will impact `jhipster-online` for the next release, that's why I added a new label `jhipster-online` for you, @jdubois

**Comment by maznag** [comment]
@jdubois same for the merged PR, don't forget to disable 'Clustered HTTP sessions using Hazelcast' option when oAuth2.0/OIDC is selected on JHipster Online

**Comment by jdubois** [comment]
Thanks @maznag it's already done :-)

File: generators/languages/templates/src/main/webapp/i18n/fr/_logs.json [hunk 5]
```diff
@@ -8,6 +8,7 @@
             "level": "Niveau"
         }
     },
+    <%_ if (databaseType !== 'no' && databaseType !== 'cassandra') { _%>
```
Comment (deepu105) [inline generators/languages/templates/src/main/webapp/i18n/fr/_logs.json]: This file has this by mistake, as we have a dedicated json for this, hence there was no need to rename logs.json

**Comment by pascalgrimaud** [comment]
oh sorry @deepu105, I already merged it as I tested and it works fine

File: generators/languages/templates/src/main/webapp/i18n/fr/_logs.json (hunk 5 above)
Comment (deepu105) [reply generators/languages/templates/src/main/webapp/i18n/fr/_logs.json]: i'm fixing it

**Comment by yelhouti** [comment]
I can not use import-jdl anymore when using this option, any one want's to discuss this?

**Comment by pascalgrimaud** [comment]
@yelhouti : don't comment old ticket. If you have an issue, plz open a new ticket with **full details** and **full steps**, so we can reproduce.
